### ISSUE TITLE
allow `raw_exec` tasks to drop client environment variables

### ISSUE THREAD

**Issue by grabango-thampton** [hashicorp/nomad#17650]
### Nomad version
`v1.5.3`

### Operating system and Environment details

No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 20.04.5 LTS
Release:        20.04
Codename:       focal

Kernel: 5.4.0-149-generic

### Issue

When an environment variable is set in a client's artifact config, it leaks into the allocations created on that client. This is happening for `raw_exec`. I haven't tested it for other driver types.

```
client {
  artifact {
    set_environment_variables = "https_proxy"
  }
}
```

In this particular context, the `https_proxy` environment variable is set to enable a forward proxy to cache packages downloads. I only really want this for downloading packages.

However, this is leaking into `raw_exec` allocations and makes it so that any running applications refer to the proxy.

### Reproduction steps

1. Set the Nomad client configuration so that an artifact environment variable is used for go-getter.
2. Create an allocation using `raw_exec` and use exec to open up a shell and print out the environment variable.

#### Expected Result

That environment variables used to download an artifact are only used to download an artifact.

#### Actual Result

Environment variables that are useful for downloading artifacts interfere with running applications.

### Job file (if appropriate)

No job file as the job doesn't actually matter.

### Nomad Server logs (if appropriate)

N/A

### Nomad Client logs (if appropriate)

N/A

**Comment by grabango-thampton** [comment]
I did notice that the landlock mechanism in: https://github.com/hashicorp/nomad/pull/15328

Requires a more up to date kernel

**Comment by grabango-thampton** [comment]
Actually, I am setting https_proxy for the Nomad client via systemd and then when Nomad forks a process for a raw_exec it's just taking the environment variables for the child process.

I really wish you could just set an environment variable in the artifact stanza.

**Comment by tgross** [comment]
Hi @grabango-thampton so yeah as you've discovered the behavior is as expected. The process we spawn for the `artifact` block scrubs its environment by default unless `set_environment_variables` is set (ref https://developer.hashicorp.com/nomad/docs/job-specification/artifact#environment) and the `raw_exec` driver runs as the same user as the Nomad agent itself without any further isolation.

If we automatically scrubbed the environment of the `raw_exec` task that would almost certainly break a lot of people's existing jobs, so from the perspective of backwards compatibility if we were to implement this sort of thing it'd be as a new flag on the `raw_exec` [task configuration](https://developer.hashicorp.com/nomad/docs/drivers/raw_exec#task-configuration) block. I'll mark this as a feature request for roadmapping.

In the meantime, you may be able to workaround the specific issue you're having by setting `https_proxy` to `""` in the task's `env` block.

**Comment by devashishTaneja** [comment]
@tgross Can I take up this issue?

**Comment by tgross** [comment]
Sure, go for it!

**Comment by devashishTaneja** [comment]
@tgross As part of this issue we need to introduce a flag to scrub the env variables from the artefact downloader context right?

**Comment by grabango-thampton** [comment]
My use case for this feature diminished as the design of our Nomad architecture changed. Initially, there was a Squid forward proxy in front of Nomad to act as a package cache and general tunnel and this required setting `https_proxy` within each artifact stanza block.

However, Squid was swapped out for Nginx and so now a reverse proxy serves as a package cache and there is no need to set that environment variable.

This is probably still a good thing to dot Is and Ts around though.

**Comment by tgross** [comment]
> As part of this issue we need to introduce a flag to scrub the env variables for the task, right?

Right. I recommend that the configuration be similar to `artifact.set_environment_variables` so that users have the option to scrub most but not all variables. With the difference being if it's unset that we don't touch the environment at all.

> I couldn't find any logic where we copy the env from the agent, so I assume we need to force unset these env? Can you give me some pointers on where to do this?

The driver sets an environment here: [`rawexec/driver.go#L334`](https://github.com/hashicorp/nomad/blob/v1.5.6/drivers/rawexec/driver.go#L334) but then that calls into the executor that's shared between the `exec`, `java`, and `raw_exec` drivers. In the `exec`/`java` drivers the environment is getting scrubbed by libcontainer, but that doesn't happen for `raw_exec`. The tricky bit is that the shared executor is sometimes imported by third-party plugins even though it's not part of our official SDK. So we'll want to avoid making a change to `executor.Launch` that isn't backwards compatible.

**Comment by tgross** [comment]
Closed by https://github.com/hashicorp/nomad/pull/25511 which is shipping in 1.10.0
